### PR TITLE
Try using socket syscalls that accepts a single buffer to improve performance

### PR DESCRIPTION
**Pull request by tmds**
This is for benchmarking to see if using syscalls that accept a single buffer has a measurable impact on performance.

cc @lpereira @adamsitnik @stephentoub

### CODE REVIEW

**Comment by adamsitnik** [comment]
@tmds may I ask you for `System.Net.Sockets.dll` and `libSystem.Native.so` files?

**Comment by tmds** [comment]
@adamsitnik I have issue compiling with pgo optimizations enabled (clang crashes for me). I don't know how it affects performance. Here is what I compiled: [socket_syscall.tar.gz](https://github.com/dotnet/runtime/files/4630434/socket_syscall.tar.gz). It may be better if you compile it yourself.

**Comment by adamsitnik** [comment]
@tmds I've compiled your changes on my Ubuntu machine and run the benchmarks for 512 connections for x64 machines.

The results look awesome! We are reaching 1180k with JSON!

obraz

The Flamegraphs:

obraz


I've uploaded the trace to:

obraz

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 1]
```diff
@@ -1595,7 +1595,7 @@ public int Receive(IList<ArraySegment<byte>> buffers, SocketFlags socketFlags, o
             if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"SRC:{LocalEndPoint} DST:{RemoteEndPoint}");
 
             int bytesTransferred;
-            errorCode = SocketPal.Receive(_handle, buffers, ref socketFlags, out bytesTransferred);
+            errorCode = SocketPal.Receive(_handle, buffers, socketFlags, out bytesTransferred);
```
Comment (tmds) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1598]: This didn't have to be a `ref`, so I removed it.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 2]
```diff
@@ -1032,25 +1135,23 @@ public static SocketError Receive(SafeSocketHandle handle, byte[] buffer, int of
...
             if (!handle.IsNonBlocking)
             {
-                return handle.AsyncContext.Receive(new Memory<byte>(buffer, offset, count), ref socketFlags, handle.ReceiveTimeout, out bytesTransferred);
+                return handle.AsyncContext.Receive(new Memory<byte>(buffer, offset, count), socketFlags, handle.ReceiveTimeout, out bytesTransferred);
             }
 
-            int socketAddressLen = 0;
             SocketError errorCode;
-            bool completed = TryCompleteReceiveFrom(handle, new Span<byte>(buffer, offset, count), socketFlags, null, ref socketAddressLen, out bytesTransferred, out socketFlags, out errorCode);
+            bool completed = TryCompleteReceive(handle, new Span<byte>(buffer, offset, count), socketFlags, out bytesTransferred, out errorCode);
```
Comment (tmds) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs:1142]: This makes us use the single-buffer syscall also for sync calls on Socket.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 3]
```diff
@@ -1032,25 +1135,23 @@ public static SocketError Receive(SafeSocketHandle handle, byte[] buffer, int of
...
             if (!handle.IsNonBlocking)
             {
-                return handle.AsyncContext.Receive(buffer, ref socketFlags, handle.ReceiveTimeout, out bytesTransferred);
+                return handle.AsyncContext.Receive(buffer, socketFlags, handle.ReceiveTimeout, out bytesTransferred);
             }
 
-            int socketAddressLen = 0;
             SocketError errorCode;
-            bool completed = TryCompleteReceiveFrom(handle, buffer, socketFlags, null, ref socketAddressLen, out bytesTransferred, out socketFlags, out errorCode);
+            bool completed = TryCompleteReceive(handle, buffer, socketFlags, out bytesTransferred, out errorCode);
```
Comment (tmds) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs:1154]: This makes us use the single-buffer syscall also for sync calls on Socket.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 4]
```diff
@@ -96,7 +96,29 @@ public static unsafe SocketError CreateSocket(AddressFamily addressFamily, Socke
             return errorCode;
         }
 
-        private static unsafe int Receive(SafeSocketHandle socket, SocketFlags flags, Span<byte> buffer, byte[]? socketAddress, ref int socketAddressLen, out SocketFlags receivedFlags, out Interop.Error errno)
+        private static unsafe int SysReceive(SafeSocketHandle socket, SocketFlags flags, Span<byte> buffer, out Interop.Error errno)
```
Comment (tmds) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs:99]: I prefixed the private methods that call into pal with Sys to make it easier to distinguish from the public methods with the same name that are called from Socket.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 5]
```diff
@@ -479,7 +480,15 @@ protected override bool DoTryComplete(SocketAsyncContext context)
                 }
                 else
                 {
-                    return SocketPal.TryCompleteReceiveFrom(context._socket, Buffer.Span, null, Flags, SocketAddress, ref SocketAddressLen, out BytesTransferred, out ReceivedFlags, out ErrorCode);
+                    if (!SetReceivedFlags && SocketAddress == null)
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:483]: In what situation is SetReceivedFlags false but SocketAddress != null?  I'm wondering why we need both of these checks and/or if SetReceivedFlags could be initialized appropriately based on SocketAddress.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 6]
```diff
@@ -137,7 +159,33 @@ private static unsafe int Receive(SafeSocketHandle socket, SocketFlags flags, Sp
...
+            {
+                int bytesSent = 0;
+                errno = Interop.Sys.Send(
+                    socket,
+                    &b[offset],
+                    count,
+                    flags,
+                    &bytesSent);
+
+                sent = bytesSent;
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs:175]: Why do we need both sent and bytesSent?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs (hunk 5 above)
Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:483]: > In what situation is SetReceivedFlags false but SocketAddress != null?

That doesn't happen.

> I'm wondering why we need both of these checks and/or if SetReceivedFlags could be initialized appropriately based on SocketAddress.

I think we need both. Consider UDP socket, when calling `Socket.ReceiveAsync(saea)`, `SetReceivedFlags = true, SocketAddress == null`. And when calling `Socket.ReceiveFromAsync(saea)`, `SetReceivedFlags = true, SocketAddress != null`.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs (hunk 6 above)
Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs]: #fixed

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:483]: But why wouldn't you just set:
SetReceivedFlags = SocketAddress != null;
in the latter case?

Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: I've removed `SocketAddress` from the check, and added an Assert instead.

> But why wouldn't you just set:
SetReceivedFlags = SocketAddress != null;
in the latter case?

I don't understand the question/suggestion. Where would this initializer be? Currently initialization is hard-coded to true/false.

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: > Where would this initializer be? Currently initialization is hard-coded to true/false.

Where you currently have it hardcoded to true.  If the only purpose of this was to control which of these paths were taken, and if everywhere it was checked we checked both conditions, then you could just use the one flag to represent both conditions by setting it appropriately based on both.

> I've removed SocketAddress from the check, and added an Assert instead.

If the check wasn't actually necessary and an assert suffices, even better.

Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: > Where you currently have it hardcoded to true.

We want it to be `true` also `Socket.ReceiveAsync(saea)` gets called for UDP. Then `SocketAddress` is `null`. (`SocketAsyncContext.ReceiveFromAsync` gets called for `Socket.ReceiveAsync`.)

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: Either I'm not understanding or we're talking past each other.  But it doesn't matter anymore as you removed the extra check.

**Pull request merged**